### PR TITLE
feat: Add jsrepo

### PR DESCRIPTION
**Pull request by ieedan**
I am opening this up as a draft so we can discuss it. 

This currently works in this state and you can test it by running:
```bash
# init then add
pnpm dlx jsrepo init --project --repos github/ieedan/geist/tree/jsrepo

pnpm dlx jsrepo add

# or

# add with zero-config
pnpm dlx jsrepo add --repo github/ieedan/geist/tree/jsrepo
```

We have to include `utils` and `icons` in the build so that we have all local dependencies covered but I have opted to not list them when users add blocks.

The problem that I want to discuss is icons...

With the current way that we do icons everything is imported into a index.ts file and then rexported to be used everywhere else. This does work, but because of this you have to install every icon even if you are only going to use 1. 

So with that explained there are a few options:
1. Convert each icon to a `.svelte` file and import each icon individually.
2. Package the icons separately and install them as a dependency (I kinda like this one personally but it is more of a pain)

I don't think this really makes any sense to merge unless we make a change so that not every icon is installed for every component.

Edit: After https://github.com/ieedan/jsrepo/pull/244 the time it takes to actually install isn't that bad but I would still find it annoying that every icon is installed in my repo after installing `Avatar`.

### CODE REVIEW

**Comment by ieedan** [comment]
@shyakadavis any thoughts on this?

If we decided what to do I can try and implement it today!

**Comment by shyakadavis** [comment]
Hey, @ieedan I think I like the idea of installing all icons at once (If this is terrible, I am open to the alternative.) Here is my only reason; you know how Vite tries to optimize everything on the first page visit? When all icons are bundled and optimized, you can navigate to other pages without Vite trying to re-load and re-optimize on each visit. (It's done all at once; does that make sense?)

**Comment by ieedan** [comment]
> Hey, @ieedan I think I like the idea of installing all icons at once (If this is terrible, I am open to the alternative.) Here is my only reason; you know how Vite tries to optimize everything on the first page visit? When all icons are bundled and optimized, you can navigate to other pages without Vite trying to re-load and re-optimize on each visit. (It's done all at once; does that make sense?)

Yeah I have seen this with lucide it's super annoying!

**Comment by ieedan** [comment]
If that's fine with you then I think I'll take this out of draft. Give it a try and let me know what you think and if we need to hide any other components from the list!

I did add concurrency so that adding the icons isn't as brutal as it was before.

**Comment by ieedan** [comment]
Another problem that comes up with this is because many components will depend on every icon jsrepo will ask if you want to overwrite every icon almost every time you add a single component and that's pretty annoying.

Either way give it a try and let me know what you think!

**Comment by ieedan** [comment]
I guess a way around the annoying prompts for every icons is to just hold `y` or provide the `-y` flag. 

Playing around with this adding some of the components to the projects works really good outside of the icons thing.  

Just need to add some instruction for the other config files I think!

**Comment by ieedan** [comment]
Also I relocated theme-switcher so that you can actually install it

**Comment by ieedan** [comment]
Full demo: https://github.com/ieedan/geist-jsrepo

**Comment by shyakadavis** [comment]
Hey, @ieedan This is really great. Doing one last pass at testing, and I believe we should be good.

**Comment by shyakadavis** [comment]
@ieedan a request:

When adding a component through `jsrepo`, the path aliases are not retained.

i.e
```svelte
<script>
	import { cn } from '$lib/utils.js';
// becomes
	import { cn } from '../../lib/utils.js';
</script>
```
Is there no current way to retain this?

**Comment by ieedan** [comment]
1. This is because you are using relative paths under `paths`. However I don't think the alias works correctly atm because jsrepo doesn't look for the ts/jsconfig under `.svelte-kit` so this would be something that would need to be solved on the jsrepo side maybe allowing for a custom path in the config or something. Maybe even a hack just for svelte kit but I will look at this today.
2. Already have an issue open for this https://github.com/ieedan/jsrepo/issues/262 should be able to get to that today since it just requires updating the `add` command
3. Currently each path maps to a category in a registry. Right now there is no way to have standalone files without a category maybe this could change in the future but it will be a pretty big change.
If I can fix no. 1 then your paths should look like:
```json
{
	"paths" : {
		"*": "./src/lib/blocks",
		"ui": "$lib/components/ui",
		"icons": "$lib/assets/icons",
		"lib": "$lib"
	}
}
```

**Comment by ieedan** [comment]
Taking a second look I think no. 1 is working? Maybe just change your paths to use aliases?

**Comment by shyakadavis** [comment]
Let me have a look @ieedan 

P.S. Heads-up; this doesn't work.
```
pnpm dlx jsrepo add --repo github/ieedan/geist/tree/jsrepo
```
Screenshot 2024-12-06 at 13 58 07

**Comment by ieedan** [comment]
Thats the zero-config option just say yes to continue or initialize jsrepo first.

**Comment by shyakadavis** [comment]
Ah. It defaulted to "no" and my immediate assumption was that I was doing something wrong, lol

**Comment by ieedan** [comment]
Yeah I default to no because zero-config adds disable a lot of what's actually useful with jsrepo but it's nice for demos.

**Comment by shyakadavis** [comment]
@ieedan awesome, man!
1&3 are immediately resolved with
```
{
	"paths" : {
		"*": "./src/lib/blocks",
		"ui": "$lib/components/ui",
		"icons": "$lib/assets/icons",
		"lib": "$lib"
	}
}
```
I didn't use this before because my assumption was that I was providing a literal "path" and silly me didn't stop to consider that an alias also applies. (A note would be nice for future "me's" 😅 )

**Comment by ieedan** [comment]
It kinda is a literal path to know where to put your blocks it has to actually resolve each alias to it's literal path. I think we can add a note in the docs under jsrepo.json -> paths.

**Comment by shyakadavis** [comment]
@ieedan is there a way to include files & other assets (e.g. tailwind.config.ts & app.css) in the source registry

**Comment by ieedan** [comment]
> @ieedan is there a way to include files & other assets (e.g. tailwind.config.ts & app.css) in the source registry

Not currently but it may be worth exploring. I almost feel like with us having the `jsrepo-build-config.json` now that we could add a key like:
```json
{
	"configs": {
		// file name
		"tailwind.config.ts": {
			"from": "./", // where in the registry
			"to": "./" // to where in the project
		},
		"vite.config.ts": {
			"from": "./",
			"to": "./"
		},
		"app.css": {
			"from": "./src",
			"to": "./src"
		}
	}
}
```

But with configs often you want to just overwrite specific values which I can't imagine supporting for every use case. The other problem here is what if tailwind.config or vite.config are js instead of ts? Or app.css is app.pcss then you would have duplicate files. Another problem you could run into is what if I am using multiple ui libraries that both want have their own configs.

I think my original vision was that jsrepo handles adding the code and then for the config you add docs in your registry or if you really need more customization roll your own simple CLI for the config files.

I think it's a difficult problem to solve for everyone and that's probably why I haven't tried but I will have to think about it some more.

**Comment by ieedan** [comment]
image

**Comment by ieedan** [comment]
Okay no. 2 should be fixed in `1.17.6`

**Comment by ieedan** [comment]
I do think before we merge this we have to work on the instructions though. I have pretty details instructions in the demo so maybe look there and let me know what you think?

File: README.md [hunk 1]
```diff
@@ -2,14 +2,136 @@
...
+export default defineConfig({
+	plugins: [
+		sveltekit(),
++		svg()
+	]
+});
+```
+
+> [!TIP]
+> For Typescript not to complain about `file.svg?component` et al., add the following import statement to `src/app.d.ts` (or any `.d.ts` file somewhere in the path of your project where `tsc` can find it).
```
Comment (ieedan) [inline README.md:62]: I think we are missing the import statement also 'et al.' ?

**Comment by shyakadavis** [comment]
Snap. Why isn't "Tip" rendering like so?
Screenshot 2024-12-06 at 15 25 04

**Comment by ieedan** [comment]
It should be:
> [!TIP]
> tip

**Comment by ieedan** [comment]
```md
> [!TIP]
> tip
```

**Comment by ieedan** [comment]
I thought it might of been line wrapping so I had claude shorten the tip but it didn't seem to make a difference

**Comment by ieedan** [comment]
@shyakadavis So I think it's a bug with github markdown rendering. The rendering is different on PR previews for whatever reason. I just pasted this into the demo and it worked perfectly.

**Comment by ieedan** [comment]
I think this is ready though unless you have any other issues that should be addressed first!

**Review by shyakadavis — APPROVED** [review]
Thank you very much, @ieedan You're awesome.

**Comment by ieedan** [comment]
> Thank you very much, @ieedan You're awesome.

Likewise my friend

**Pull request merged**